sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1656,6 +1688,170 @@ pi_command_list_ptr_t _pi_queue::eventOpenCommandList(pi_event Event) {
   return CommandListMap.end();
 }
 
+pi_result _pi_queue::insertBarrier(pi_uint32 NumEventsInWaitList,
+                                   const pi_event *EventWaitList,
+                                   pi_event *Event) {
+  // Helper function for appending a barrier to a command list.
+  auto insertBarrierIntoCmdList =
+      [this](pi_command_list_ptr_t CmdList,
+             const _pi_ze_event_list_t &EventWaitList, pi_event &Event) {
+        if (auto Res = createEventAndAssociateQueue(
+                this, &Event, PI_COMMAND_TYPE_USER, CmdList))
+          return Res;
+        Event->WaitList = EventWaitList;
+        ZE_CALL(zeCommandListAppendBarrier,
+                (CmdList->first, Event->ZeEvent, EventWaitList.Length,
+                 EventWaitList.ZeEventList));
+        return PI_SUCCESS;
+      };
+
+  // If we have a list of events to make the barrier from, then we can create a
+  // barrier on these and use the resulting event as our future barrier.
+  if (NumEventsInWaitList) {
+    // Get an arbitrary command-list in the queue.
+    pi_command_list_ptr_t CmdList;
+    if (auto Res = Context->getAvailableCommandList(this, CmdList, false))
```

Comment:
Any reason for this and other new calls to `getAvailableCommandList` to not allow batching?

Reply:
I have added a small clarification of what we need them for. Is that sufficient?

Reply:
> My primary concern was that batching could cause issues when a barrier is inserted into a batch on compute, followed by an operation with a copy using the active barrier and getting stuck waiting on the unsubmitted batch. I am not too familiar with the batch mechanism, so maybe that cannot happen? If not, then I am happy to go enable batching.

I think the "barrier" is not any specific for copy vs. compute forward-progress. One thing I recall is that I specifically needed to allow batching for barriers, see here: https://github.com/intel/llvm/pull/5307
It was causing seeing some slowdowns without batching.

Reply:
@steffenlarsen : are you going to enable batching for these to avoid performance regressions?

Reply:
Batching has been enabled. There is one exception to this, specifically if there are multiple queues when looking up one command-list for each queue. There is a comment for why, but in short the continuous lookups will push out the batching command lists.